src/native/managed/cdacreader/Microsoft.Diagnostics.DataContractReader.Abstractions/Contracts/IRegistry.cs

```diff
@@ -0,0 +1,18 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.Diagnostics.DataContractReader.Contracts;
+
+internal interface IRegistry
```

Comment:
I assume this is a rhetorical question asking me to document this interface? If so, done.  If not, i'm open to suggestions for how to make it clear what this is. At some point we decided that contracts for managed objects will use the managed class name as the contract name. We can revisit that if you like

Reply:
It is entirely possible I missed some design conversation. I don't recall what the `IRegistry` is being used for though. Part of my confusion is that Exception is on the same level as Object and Thread. It seems like the three of them are all instances whereas DacStreams would be more a global concept. I'm simply not understanding the utility of this interface.

Reply:
Ah ok. So your question is "we have `ITarget`", why can't it just specify that `ITarget.Contracts` returns a concrete class like this:

```csharp
interface ITarget
{
    ContractRegistry Contracts { get; }

    public class ContractRegistry
    {
        public IException Exception => ... ; // implementation that uses the ExceptionFactory
        ... // and so on for the other contracts
    }
}
```

Why do we instead introduce an interface here?

There are two motivations for having an interface here:

1. It makes it possible to implement new contracts by coding against the interfaces defined in `Microsoft.Diagnostics.DataContracts.Abstractions.dll` (for example for some of our future plans we will be able to keep some implementations in a separate assembly that we bring in during the final NativeAOT publish step) rather than depending on the concrete implementation of all the other contract factories and contracts.

2. It makes it possible to separately test contract implementations by providing an `IContractRegistry` implementation that returns mock objects in unit tests instead of depending on a real contract implementation.  (This is my main motivation currently - for example the validation logic for `RuntimeTypeSystem.MethodDescHandle` in (https://github.com/dotnet/runtime/pull/106413) depends on the CodeVersions and PrecodeStubs contracts which access a ton of data structures in the target process.  For unit testing the MethodDescHandle validation I could instead just depend on mock instances of CodeVersions and Precode stubs that just return predefined answers and don't depend on anything)

Reply:
Wrt use of interfaces: I think that the .NET API design guidelines would prefer abstract classes in number of situations where we have used interfaces. Is there a reason why we are not using abstract classes? 

(I am not sure how much we care about following .NET API design guidelines.)

Reply:
> would prefer abstract classes in number of situations where we have used interfaces. Is there a reason why we are not using abstract classes?

I think it wasn't so much that we prefer to use interfaces instead of abstract classes, it's that we wanted to use structs instead of classes for the contracts.  I think the idea was to emphasize that contract instances have no identity and are morally are stateless - all state is in the `Target`.  That said, we've do have some cached state: the various `XyzHandle` values are used be the contracts as keys into caches to avoid repeated validation.

Original discussion: https://github.com/dotnet/runtime/pull/101994#pullrequestreview-2048264913

Reply:
On the other hand, specifically for `ITarget`/`IContractRegistry` I don't have a particular problem with doing abstract classes instead: `AbstractTarget` / `AbstractContractRegistry` something like that.  I think the two motivating scenarios will work just as well.  I'll push an update

Reply:
updated

* abstract class `Target`
* concrete implementation: `ContractDescriptorTarget` - populates itself based on a contract descriptor
* abstract class `AbstractContractRegistry`
* concrete implementation: `ContractRegistry` - creates contract instances using a collection of `IContractFactory<TContract>` instances

Reply:
> AbstractContractRegistry

Nit: I do not think we have any public API abstraction that starts with `Abstract`. The usual naming convention is to name the abstraction suffix, and make the name of concrete implementation more specific, e.g. `Stream` vs. `FileStream`.